src/libsystemd/sd-device/device-monitor.c

```diff
@@ -450,6 +456,35 @@ static int passes_filter(sd_device_monitor *m, sd_device *device) {
         return device_match_parent(device, m->match_parent_filter, m->nomatch_parent_filter);
 }
 
+static bool check_sender_uid(sd_device_monitor *m, uid_t uid, bool is_unicast) {
+        int r;
+
+        assert(m);
+
+        /* Always trust messages from uid 0. */
+        if (uid == 0)
+                return true;
+
+        if (!m->uid_range_loaded) {
+                r = uid_range_load_userns(&m->mapped_userns_uid_range, &m->n_uid_range, NULL);
+                if (r < 0)
+                        log_monitor_errno(m, r, "Failed to load UID ranges mapped to the current user namespace, ignoring: %m");
+                else
+                        m->uid_range_loaded = true;
+        }
+
+        /* Trust messages come from outside of the current user namespace. */
+        if (m->uid_range_loaded && !uid_range_contains(m->mapped_userns_uid_range, m->n_uid_range, uid))
+                return true;
+
+        /* Trust unicast messages sent by the same UID we are running. */
+        if (is_unicast && (uid == getuid() || uid == geteuid()))
+                return true;
```

Comment:
i'd move this check above the userns check, given it's simple and quick, and userns might not be